### ISSUE TITLE
Difference between the CPU time and the GPU time?

### ISSUE THREAD

**Issue by MoonJian** [awwong1/torchprof#5]
Thanks for your helpful tools! I still have a question about the difference between the CPU time and the GPU time. I deploy CNN models on GPU platform, if I want to measure the latency of each layer, can I directly use the cuda total time as the results? Does the CPU total time include the cuda total time as explained by the given link [Self CPU Time vs CPU Time](https://software.intel.com/en-us/vtune-amplifier-help-self-time-and-total-time)? Thank you!

**Comment by awwong1** [comment]
CPU times are separate from the CUDA times. You have some amount of CPU overhead in addition to  the amount of time it took for the GPU to run your model.

For each layer, you should sum the `Self CPU total` and `CUDA total` columns.

**Comment by soumendukrg** [comment]
I have a related question @awwong1 ,  don't you think to account for all CPU overhead, we should use `CPU total` + `CUDA total` instead of `Self CPU total` + `CUDA total`?

**Comment by awwong1** [comment]
Hi @soumendukrg, I think the answer to this question depends. I simply return the raw pytorch profiler results and allow the researcher to do the aggregations as appropriate to their study.
Please also refer to the explanation I write here: https://github.com/awwong1/torchprof/issues/6#issuecomment-582120080

**Comment by soumendukrg** [comment]
I did refer to the comment you mentioned and seems like `self.cpu_time_total` measures the start time to end time and it already considers how much time any of the children takes. Please correct me if I am wrong.

On another note, let me rephrase my earlier question:
I want to find out layer-wise execution time of AlexNet on **CPU** only.
So, to find out how much time a particular layer takes to execute, which one is more appropriate: `self cpu total` or `cpu total`?

I use something like this:

```python
model = torchvision.models.alexnet()
inp = torch.rand((1, 3, 224, 224))

start = time.time()
with torchprof.Profile(model, use_cuda=False, paths=paths) as prof:
       model(inp)
end = time.time()
```
Now if I add up all layer times from the column `self cpu total`, it closely matches `end - start`, while adding all times from column `cpu total` gives much more.

Could you comment on this behavior?

**Comment by awwong1** [comment]
>  don't you think to account for all CPU overhead, we should use CPU total + CUDA total instead of Self CPU total + CUDA total?

Sorry, I did not understand your original question. You should not use CUDA if you only care about CPU.

To find out how much time a layer takes to execute, `Self CPU total` is the correct metric.

----
Example:

```python
import torch
import torchvision
import torchprof

model = torchvision.models.alexnet(pretrained=False)
x = torch.rand([1, 3, 224, 224])

with torchprof.Profile(model, use_cuda=False) as prof:
    model(x)

print(prof.display(show_events=False))
```
```text
Module         | Self CPU total | CPU total | CUDA total | Occurrences
---------------|----------------|-----------|------------|------------
AlexNet        |                |           |            |            
├── features   |                |           |            |            
│├── 0         |       16.939ms |  67.654ms |    0.000us |           1
│├── 1         |      130.908us | 212.401us |    0.000us |           1
│├── 2         |        3.721ms |   7.456ms |    0.000us |           1
│├── 3         |        5.615ms |  22.410ms |    0.000us |           1
│├── 4         |      105.680us | 168.074us |    0.000us |           1
│├── 5         |        2.729ms |   5.474ms |    0.000us |           1
│├── 6         |        3.893ms |  15.519ms |    0.000us |           1
│├── 7         |       87.431us | 133.692us |    0.000us |           1
│├── 8         |        5.030ms |  20.062ms |    0.000us |           1
│├── 9         |       95.478us | 141.983us |    0.000us |           1
│├── 10        |        3.465ms |  13.807ms |    0.000us |           1
│├── 11        |       99.031us | 149.575us |    0.000us |           1
│└── 12        |      887.170us |   1.794ms |    0.000us |           1
├── avgpool    |        2.239ms |   4.495ms |    0.000us |           1
└── classifier |                |           |            |            
 ├── 0         |      173.451us | 433.070us |    0.000us |           1
 ├── 1         |        9.896ms |   9.918ms |    0.000us |           1
 ├── 2         |       68.335us |  94.130us |    0.000us |           1
 ├── 3         |      213.936us | 533.142us |    0.000us |           1
 ├── 4         |        5.425ms |   5.454ms |    0.000us |           1
 ├── 5         |       73.571us | 101.251us |    0.000us |           1
 └── 6         |        1.586ms |   1.620ms |    0.000us |           1
```
```python
trace, event_lists_dict = prof.raw()
print(event_lists_dict[('AlexNet', 'features', '0')][0])
```
```text
----------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------------------------------------  
Name                    Self CPU total %  Self CPU total   CPU total %      CPU total        CPU time avg     Number of Calls  Input Shapes                                   
----------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------------------------------------  
conv2d                  0.07%            12.494us         100.00%          16.939ms         16.939ms         1                []                                             
convolution             0.06%            10.868us         99.93%           16.926ms         16.926ms         1                []                                             
_convolution            0.40%            67.620us         99.86%           16.915ms         16.915ms         1                []                                             
size                    0.02%            3.400us          0.02%            3.400us          3.400us          1                []                                             
size                    0.01%            0.875us          0.01%            0.875us          0.875us          1                []                                             
size                    0.00%            0.744us          0.00%            0.744us          0.744us          1                []                                             
size                    0.00%            0.704us          0.00%            0.704us          0.704us          1                []                                             
size                    0.01%            0.869us          0.01%            0.869us          0.869us          1                []                                             
size                    0.01%            0.875us          0.01%            0.875us          0.875us          1                []                                             
size                    0.00%            0.843us          0.00%            0.843us          0.843us          1                []                                             
contiguous              0.02%            4.001us          0.02%            4.001us          4.001us          1                []                                             
contiguous              0.01%            1.154us          0.01%            1.154us          1.154us          1                []                                             
contiguous              0.01%            1.130us          0.01%            1.130us          1.130us          1                []                                             
mkldnn_convolution      99.22%           16.807ms         99.38%           16.833ms         16.833ms         1                []                                             
empty                   0.12%            21.021us         0.12%            21.021us         21.021us         1                []                                             
as_strided_             0.03%            4.783us          0.03%            4.783us          4.783us          1                []                                             
----------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------------------------------------  
Self CPU time total: 16.939ms
```

**Comment by soumendukrg** [comment]
Thanks for the response. I understand that if I am not using CUDA, I should not care about that as all the times are already 0.

**Comment by soumendukrg** [comment]
In the same way, if I want to find out how much time a layer takes to execute in a machine which has GPU, what is the correct metric?

My code looks like this:
```python
device = 'cuda:0'
model = torchvision.models.alexnet()
model.to(device)
inp = torch.rand((1, 3, 224, 224))
inp.to(device)

start = time.time()
with torchprof.Profile(model, use_cuda=True, paths=paths) as prof:
       model(inp)
end = time.time()
```

**Comment by awwong1** [comment]
There is no GPU equivalent metric `self_cuda_time_total`.
You should emit the nvtx range and use nvvp instead.

**Comment by egamezl** [comment]
> CPU times are separate from the CUDA times. You have some amount of CPU overhead in addition to the amount of time it took for the GPU to run your model.
> 
> For each layer, you should sum the `Self CPU total` and `CUDA total` columns.

Hi, what do you mean with CPU **overhead**? I am doing something similar as the original author´s question but i want to understand the difference between the CPU and GPU times. From you answer seems like the GPU runs the AI model, so like the identification of objects and such, but what does the CPU does? Thanks in advance.

**Comment by awwong1** [comment]
> Hi, what do you mean with CPU **overhead**?

There will always be some CPU computation even if the model is configured to use CUDA.

As for the specifics, it depends on your model. I suggest looking at the official torch.profiler [profile.export_chrome_trace](https://pytorch.org/docs/stable/profiler.html#torch.profiler.profile.export_chrome_trace) method to get a more visual insight (or you can step through the events manually).
